sdk/identity/identity/src/credentials/deviceCodeCredential.ts

```diff
@@ -0,0 +1,180 @@
+        "Content-Type": "application/x-www-form-urlencoded"
+      },
+      abortSignal: options && options.abortSignal
+    });
+
+    const response = await this.identityClient.sendRequest(webResource);
+    if (!(response.status === 200 || response.status === 201)) {
+      throw new AuthenticationError(response.status, response.bodyAsText);
+    }
+
+    return response.parsedBody as DeviceCodeResponse;
+  }
+
+  private async pollForToken(
+    deviceCodeResponse: DeviceCodeResponse,
+    options?: GetTokenOptions
+  ): Promise<AccessToken | null> {
+    let accessToken: AccessToken | null = null;
+
+    const webResource = this.identityClient.createWebResource({
+      url: `${this.identityClient.authorityHost}/${this.tenantId}/oauth2/v2.0/token`,
+      method: "POST",
+      disableJsonStringifyOnBody: true,
+      deserializationMapper: undefined,
+      body: qs.stringify({
+        grant_type: "urn:ietf:params:oauth:grant-type:device_code",
+        client_id: this.clientId,
+        device_code: deviceCodeResponse.device_code
+      }),
+      headers: {
+        Accept: "application/json",
+        "Content-Type": "application/x-www-form-urlencoded"
+      },
+      abortSignal: options && options.abortSignal
+    });
+
+    while (accessToken === null) {
+      try {
+        await delay(deviceCodeResponse.interval * 1000);
+
+        // Check the abort signal before sending the request
+        if (options && options.abortSignal && options.abortSignal.aborted) {
+          return null;
+        }
+
+        accessToken = await this.identityClient.sendTokenRequest(webResource);
+      } catch (err) {
+        if (err instanceof AuthenticationError) {
+          switch (err.errorResponse.error) {
+            case "authorization_pending":
+              break;
+            case "authorization_declined":
+              return null;
+            case "expired_token":
+              throw err;
+            case "bad_verification_code":
+              throw err;
+          }
+        } else {
+          throw err;
+        }
+      }
+    }
+
+    return accessToken;
+  }
+
+  /**
+   * Authenticates with Azure Active Directory and returns an {@link AccessToken} if
+   * successful.  If authentication cannot be performed at this time, this method may
+   * return null.  If an error occurs during authentication, an {@link AuthenticationError}
+   * containing failure details will be thrown.
+   *
+   * @param scopes The list of scopes for which the token will have access.
+   * @param options The options used to configure any requests this
+   *                TokenCredential implementation might make.
+   */
+  public async getToken(
+    scopes: string | string[],
+    options?: GetTokenOptions
+  ): Promise<AccessToken | null> {
+    const scopeString = typeof scopes === "string" ? scopes : scopes.join(" ");
+    const deviceCodeResponse = await this.sendDeviceCodeRequest(scopeString, options);
+
+    this.userPromptCallback({
```

Comment:
This is a traditional callback, though I wonder if we should use something more promise-y here. I'm not entirely sure if https://azuresdkspecs.z5.web.core.windows.net/TypeScriptSpec.html#ts-use-promises pertains to this kind of case, but figured I'd mention it.

Reply:
This is meant to be a simple callback to the application code that enables it to display the device code authorization message.  We don't really care what the application does with it and don't wait on it to return, so I think a simple callback is the best approach at the moment.

However, if folks can think of other scenarios where a Promise (or a Promise-returning callback) makes more sense, I am definitely willing to change it!

Reply:
Sounds good. Just confirming with Brian (but it's fine with me). 

@bterlson - are callbacks good for fire-and-forget?

Reply:
I'm not sure that you want this to be fire or forget necessarily.  What if an exception is thrown from the users callback?  Shouldn't that be bubbled up back to the user somehow?

Reply:
The current implementation will allow an exception in the callback to be bubbled up to the SDK call, I'm invoking the callback directly in the code path without any try/catch around it.

Reply:
****

Reply:
Chatted with @bterlson about the callback design here and he agreed that this is sufficient for now.  A promise can't be used because the application code needs to be notified anytime the device code authorization flow is initiated, which could be at any time and more than once during the life of the application.

If we can collectively think of a better design for this after this PR gets merged, I'll be happy to send a new PR to improve it.